Prometheus: Fuzzy search for metric names in Code Mode

NWRichmond opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

## What is this feature?

A performance boost for the Prometheus data source's [Code Mode][code-mode], when a Prometheus instance has hundreds of thousands or millions of metric names.

### How it works

In Code Mode, if there are more than _n_ metric names (where `n=10000` by default, configurable by the user), we no longer attempt to provide the Monaco Editor with all metric names as possible autocomplete suggestions. Instead, we use fuzzy search (powered by [`@leeoniya/ufuzzy`](https://github.com/leeoniya/uFuzzy)) to provide only the _n_-most relevant metric names to the Monaco editor as possible autocomplete suggestions.

#### Performance

The resulting performance boost is very noticeable as the number of metric names gets into the hundreds of thousands. With a casual, unscientific test with 750,000 metric names, I observed an ~89% decrease in the longest-running [tasks](https://web.dev/articles/optimize-long-tasks) while typing in the code editor:

<details>
  <summary>Performance test: the "before" state</summary>

  When typing, [tasks](https://web.dev/articles/optimize-long-tasks#what_is_a_task_what-is-tasks) took between 1080-2070ms.

![suggestions perf before fix](https://github.com/grafana/grafana/assets/5732000/33d1f637-83a0-4ca7-a199-f14734bf16b6)
</details>

<details>
  <summary>Performance test: with fuzzy search</summary>

  When typing, [tasks](https://web.dev/articles/optimize-long-tasks#what_is_a_task_what-is-tasks) took between 130-203ms.

![suggestions perf after fix](https://github.com/grafana/grafana/assets/5732000/784733d9-8a05-4dbf-b9c6-aa690c89e794)
</details>

#### UI changes

The metric names fuzzy search is configured to prioritize rendering performance and single typos. Consequently, when search is active, the autocomplete will not include some suggestions that Monaco would normally display. For that reason, this PR introduces a disclaimer about the autocomplete results when metric names fuzzy search is active. The disclaimer UI only appears when metric names are in excess of the configured threshold:

https://github.com/grafana/grafana/assets/5732000/5efa88c1-bd13-4ce3-9eff-ba2c52fe3102

This PR also adds a new configuration parameter in the **Performance** section. It allows users to configure the number of metric names suggestions in Code Mode if the 10,000 default doesn't suit their needs:

https://github.com/grafana/grafana/assets/5732000/1a7882a5-b713-46e4-af0c-fa58cab9856f

## Why do we need this feature?

As described in #84612, when a Prometheus instance has an enormous number of unique metric names (on the order of many hundreds of thousands, or millions), it takes a long time (single to double-digit seconds) to load metric names to power the autocomplete experience in [Code Mode][code-mode]. Rendering such a large volume of label names in the Monaco editor can freeze the UI.

## Who is this feature for?

Prometheus data source users whose Prometheus instances have hundreds of thousands (or more) of metric names.

## Which issue(s) does this PR fix?

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #84612 

## Special notes for your reviewer

- This is my first experience with creating a new feature toggle. I'd appreciate extra focus on the implementation of the `prometheusCodeModeMetricNamesSearch`.
- To simplify the review process, I've only made Prometheus data source changes in  `packages/grafana-prometheus`. These changes will also need to be copied over to `public/app/plugins/datasource/prometheus`. I'll do that after we're aligned on the changes.
- If running this locally, you'll need to enable two feature toggles in `custom.ini`: 
  - `usePrometheusFrontendPackage`
  - `prometheusCodeModeMetricNamesSearch`

## To-do before merging

- [x] Copy updates from `packages/grafana-prometheus` to `public/app/plugins/datasource/prometheus`

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

[code-mode]: https://grafana.com/docs/grafana/latest/datasources/prometheus/query-editor/#code-mode

NWRichmond committed:
perf: limit metric names in Code Mode suggestions

NWRichmond committed:
feat: configurable metric names limit

NWRichmond committed:
feat: code mode autocomplete info/disclaimer

NWRichmond committed:
chore: put new functionality behind new feature toggle

NWRichmond commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/data-provider.ts:
This is a refactor of the `dataProvider` that was in `components/monaco-query-field/MonacoQueryField.tsx`

NWRichmond committed:
refactor: avoid type assertions

NWRichmond committed:
refactor: avoid explicit `any`

refactor: type guards

NWRichmond committed:
refactor: type guards

NWRichmond committed:
chore: add testdata results

NWRichmond committed:
fix: add missing feature toggle guard

NWRichmond commented on pkg/services/store/testdata/public_testdata.golden.jsonc:
I'm not sure why this differs from `main`

NWRichmond committed:
perf: prefer array access to `Array.prototype.at`

NWRichmond committed:
test: add missing config override

NWRichmond committed:
test: refactor for brevity & clarity

NWRichmond commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -16,25 +22,45 @@ type Completion = {
 
-// we order items like: history, functions, metrics
+  for (const idx of idxs ?? []) {
+    const metric = haystack.at(idx);
```
For curious minds: array access is ~11% faster than `Array.prototype.at`:

![jsperf-array-access-vs-array-prototype-at](https://github.com/grafana/grafana/assets/5732000/e303c928-cb5a-4190-ba98-344a35317c94)

leeoniya replied:
what does this look like with 10k values and with a `for` loop instead of forEach callback?

leeoniya replied:
thanks!

good to know. last time i checked .at() was like 50% slower. maybe it's catching up :)

generally when iterating possibly thousands of values, i try to stick with for loops and direct access since they're the most optimized in js VMs.

NWRichmond committed:
perf: avoid unnecessary mapping

NWRichmond commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -16,25 +22,45 @@ type Completion = {
-};
+const metricNamesSearchClient = new UFuzzy({
+  intraMode: 1,
+  intraChars: '[a-z0-9_:]', // characters allowed in metric names, according to the Prometheus data model
```
I just did some local testing...looks like using the default `intraChars` setting changes the results.

I created an array of metric names containing valid Prometheus characters like `_` and `:`, e.g.:
```js
["metric_0_tes:ting_delimiters","metric_1_tes:ting_delimiters",...]
```

When I use the default `intraChars`, I'm getting no matches on needles `gdeli` or `sting`. But those needles return matches when using `intraChars: '[a-z0-9_:]'`.

Here are the findings (note the `filteredMetricNames.length`):

### With default `intraChars` (`[a-z\d']`)

Needle: `gdeli`

![default-intrachars-input-gdeli-filteredMetricNames-length-0](https://github.com/grafana/grafana/assets/5732000/f715d6da-d14a-4295-abdd-6a3db7806c72)

Needle: `sting`

![default-intrachars-input-sting-filteredMetricNames-length-0](https://github.com/grafana/grafana/assets/5732000/fda1261a-31d5-4c57-b2af-ee9f3e6c16f9)

### With `intraChars: '[a-z0-9_:]'

Needle: `gdeli`

![prom-data-model-intrachars-input-gdeli-filteredMetricNames-length-10k](https://github.com/grafana/grafana/assets/5732000/b82dee05-bd2c-48e6-a8bc-30cd5aff97a1)

Needle: `sting`

![prom-data-model-intrachars-input-sting-filteredMetricNames-length-10k](https://github.com/grafana/grafana/assets/5732000/a73a84e5-89d1-4fb8-a485-2efc703ccb74)

NWRichmond replied:
Per external discussion, it seems that in order to optimize the user experience (by having minimal changes to autocomplete results when metric name search is active), the key is to operate in [`intraMode: 0`, with `intraIns: Infinity` and `intraChars: '.'`](https://github.com/leeoniya/uFuzzy?tab=readme-ov-file#options).

Many thanks, @leeoniya!

NWRichmond replied:
In https://github.com/grafana/grafana/pull/85396/commits/5428287c8891045814ba5c6e704f7265f04d9b81, I updated the settings to match option 2 above after comparing the performance between the two options. While option 1 did a great job at not changing the autocomplete results when search is active, it's almost twice as slow with large metric name arrays (tested 50k up to 2M). So until users demand closer matches, let's err on the side of performance and a fluid user experience.

NWRichmond committed:
chore: undo testdata changes

NWRichmond committed:
fix: use correct limit; perf optimizations

NWRichmond committed:
refactor: avoid unnecessary `async`s

NWRichmond committed:
types: simplify

NWRichmond committed:
test: add missing tests

NWRichmond committed:
fix: avoid hardcoding

NWRichmond committed:
test: update mock path

NWRichmond committed:
docs: fix typo

style: remove formatting artifact

style: remove formatting artifact

style: remove formatting artifact

bohandley commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/data-provider.ts:
```diff
@@ -0,0 +1,117 @@
+        detail: { limit: this.metricNamesSuggestionLimit, datasourceUid: this.languageProvider.datasource.uid },
+      })
+    );
+  }
```
Running locally, I'm having trouble seeing the listener catch this event in the component `PromQueryCodeEditorAutocompleteInfo.tsx` that shows the warning. I might be missing something, but having trouble following the logic to get the warning to show up

NWRichmond replied:
Sorry about that, @bohandley! There was a regression. Fixed now, by https://github.com/grafana/grafana/pull/85396/commits/49f8ce5bcfd4aef528a8c490c32e06b5b8373ce7:

https://github.com/grafana/grafana/assets/5732000/97a52405-c9ff-442c-ab61-7958801bea3d

bohandley replied:
Beautiful! Works like a charm. Fantastic work on this!!!

catherineymgui commented:
From a UX perspective:
- Would shorten the tooltip after "autocomplete suggestions limited" to

> The number of metric names exceeds the autocomplete limit. Only the n-most relevant metrics are displayed. You can adjust the threshold in the data source settings.

- Would put the info icon after the text

- In the configuration page - update tooltip to
> The maximum number of metric names that may appear as autocomplete suggestions in the query editor's Code mode.

NWRichmond committed:
fix: event scope regression

NWRichmond committed:
style: refactor for clarity

bohandley reviewed (approved):
Excellent work! 🚀🚀🚀

itsmylife commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -176,19 +187,19 @@ async function getLabelValuesForMetricCompletions(
-      return [...historyCompletions, ...FUNCTION_COMPLETIONS, ...metricNames];
+      const metricNames = getAllMetricNamesCompletions(dataProvider);
+      const historyCompletions = getAllHistoryCompletions(dataProvider);
+      return Promise.resolve([...historyCompletions, ...FUNCTION_COMPLETIONS, ...metricNames]);
```
Do you think we should use `Promise.all()` here?

NWRichmond replied:
`getAllMetricNamesCompletions` and `getAllHistoryCompletions` used to be `async`, but aren't `async` anymore and do not return promises.

itsmylife commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/data-provider.ts:
```diff
@@ -0,0 +1,117 @@
+import { HistoryItem } from '@grafana/data';
+// @ts-ignore
+import type { Monaco } from '@grafana/ui';
```
Do we need this? What happens when we remove this import?

NWRichmond replied:
If we remove this, it breaks the TSDoc [`@link`](https://tsdoc.org/pages/tags/link) below.

With Monaco type import:
![tsdoc link Monaco Range](https://github.com/grafana/grafana/assets/5732000/5db90b74-84ba-47ee-b737-2ac9eed8d650)

Without:
![broken tsdoc link Monaco Range](https://github.com/grafana/grafana/assets/5732000/24e17f90-5bfa-4186-81cd-cf82ea20d08d)

If you feel the `@link` is of questionable value, I'm happy to remove it.

itsmylife replied:
Do you mind commenting on this over there? It will help us in the future I believe.

NWRichmond replied:
Certainly! Updated in https://github.com/grafana/grafana/pull/85396/commits/7b9510c29586c8402cd87872b99f5d580d66e44e.

itsmylife commented on packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/data-provider.ts:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  getHistory(): string[] {
+    return this.historyProvider.map((h) => h.query.expr).filter((expr) => expr !== undefined);
```
We don't check against `null`. Do you think we should use `filter(Boolean)`?

Additionally, can't we do filtering without mapping?: 
```
return this.historyProvider.filter((h) => h.query.expr !== undefined);
```

NWRichmond replied:
> Do you think we should use filter(Boolean)?

Totally! Updated in 91c272c3fd0e025af3904c2a13ba38e9afe439b1.

> Additionally, can't we do filtering without mapping?
> ```
> return this.historyProvider.filter((h) => h.query.expr !== undefined);
> ```

We still need to map to get the `query.expr` from each history item (return type should be `string[]` and not `HistoryItem<PromQuery>[]`).

itsmylife commented on packages/grafana-prometheus/src/querybuilder/components/PromQueryCodeEditorAutocompleteInfo.tsx:
```diff
@@ -0,0 +1,67 @@
+
+    addEventListener(CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT, handleSuggestionsIncompleteEvent);
+
+    return () => {
+      removeEventListener(CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT, handleSuggestionsIncompleteEvent);
+    };
```
❤️

itsmylife commented on packages/grafana-prometheus/src/querybuilder/components/PromQueryCodeEditorAutocompleteInfo.tsx:
```diff
@@ -0,0 +1,67 @@
+      }
+    };
+
+    addEventListener(CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT, handleSuggestionsIncompleteEvent);
```
I know the datasourceUid is not gonna change and if it changes, we will be in a new datasource. But is that a safe practice to have the event handler in `useEffect`? The feature is working well and not saying we should change this but just want to question if that's safe :)

NWRichmond replied:
Probably best to have it in a `useCallback`, since it's relying on a prop.

Updated in 065f2bd097c3db55a09a8ede82088ffd0a2973d3 :)

itsmylife reviewed (approved):
Looks fantastic 🚀 
I just added a couple of questions. This is good to go!

NWRichmond committed:
refactor: prefer `useCallback` to in-effect handler

NWRichmond committed:
refactor: simplify & broaden `filter`

NWRichmond committed:
refactor: rename file to keep with conventions

NWRichmond committed:
chore: mirror Prometheus package changes in app

NWRichmond committed:
refactor: prefer no `@ts-ignore`

NWRichmond committed:
chore: update betterer results

NWRichmond committed:
docs: use type in TSDoc `@link` without `@ts-ignore`

NWRichmond committed:
test: add missing provider

NWRichmond committed:
test: fix jest mock path

NWRichmond committed:
fix: display disclaimer in empty input case
